### DISCUSSION TITLE
How to access input variables of designed workflows called as activity?

### DISCUSSION THREAD

**Discussion by JochenSteinbrenner** [elsa-workflows/elsa-core#4263] [Q&A]
Hello all,
I want to create workflows using the designer which can be called directly or as activity. 
The workflows are called using the method `workflowRuntime.StartWorkflowAsync`
with defined input parameters. 

If directly called, I can access the input parameters via
javascript `getInput('param1')`.

If the workflow is marked as "useable as activty" and I call the workflow within a parent workflow
(by setting the input parameter 'param1' within the activity) this is no longer possible.
The child workflow excepts with `getInput('param1')` not present within dictionary.
The only possible way I found, is to use `getVariable('param1')` method instead, but
this leads an exception, if the workflow is directly called.

Do you have any suggestions?

with kind regards
Jochen Steinbrenner

**Comment by sfmskywalker** [comment]
Hi @JochenSteinbrenner,

I understand your query.

To create the workflow you're aiming for:
- Start by defining Inputs and Outputs in the Input/Output tab.
- Utilize the SetOutput activity to assign a value to a specific outcome.

For clarity, I've crafted a Sum workflow as an example. It's versatile and can be used directly (like via a REST API endpoint) or as an activity:
[sum-wf.json.zip](https://github.com/elsa-workflows/elsa-core/files/12566610/sum-wf.json.zip)

After importing and publishing it, you can call it like this (ensure you use the right definition ID and API key or bearer token):

```bash
curl --location 'https://localhost:5001/elsa/api/workflow-definitions/7e7e67e5238b4d81a9e1eafdbf0b7254/execute' \
--header 'Content-Type: application/json' \
--header 'Authorization: ApiKey {secret}' \
--data '{
    "input": {
        "A": 3,
        "B": 7
    }
}'
```

Note: The Sum workflow employs JS functions such as `getA()` and `getB()`. These are dynamically generated based on the workflow's input definitions.

The output should resemble:

```json
{
    "workflowState": {
        "id": "10ed6763d79c43c2a1630db5e84b6a0e",
        "definitionId": "7e7e67e5238b4d81a9e1eafdbf0b7254",
        "definitionVersionId": "54d2cd60d24e4786823a88d28e96cdf2",
        "definitionVersion": 3,
        "status": "Finished",
        "subStatus": "Finished",
        "bookmarks": [],
        "completionCallbacks": [],
        "activityExecutionContexts": [],
        "executionLogSequence": 8,
        "input": {
            "A": 3,
            "B": 7
        },
        "output": {
            "TheResult": 10
        },
        "properties": {
            "PersistentVariablesDictionary": {
                "10ed6763d79c43c2a1630db5e84b6a0e:Workflow1:JsResult": 10
            }
        },
        "createdAt": "2023-09-09T20:12:08.831769+00:00",
        "updatedAt": "2023-09-09T20:12:08.851203+00:00",
        "finishedAt": "2023-09-09T20:12:08.851203+00:00"
    }
}
```

Additionally, I've made a Calculator workflow that showcases the Sum workflow in action:
[calculator-wf.json.zip](https://github.com/elsa-workflows/elsa-core/files/12566616/calculator-wf.json.zip)

You can run this directly from the designer:
![2023-09-09_22-20-49 (1)](https://github.com/elsa-workflows/elsa-core/assets/938393/2719fac1-e18a-42bc-a4a6-9198de2a34b5)

Hope this clarifies things! Reach out if you have more questions.
